### ISSUE TITLE
FA4 backward does not return against cutlass-dsl 4.6.0 and 4.6.1 on B300 (sm_103)

### ISSUE THREAD

**Issue by andylizf** [Dao-AILab/flash-attention#2797]
On a B300 (sm_103), FA4's backward stops returning when the loaded CuTe DSL runtime is 4.6.0 or 4.6.1. The pinned prerelease and the current release are both fine.

| nvidia-cutlass-dsl runtime | dense backward | varlen backward |
|---|---|---|
| 4.6.0.dev0 (the pin) | completes | completes |
| 4.6.0 | no return in 180s | completes |
| 4.6.1 | no return in 180s | completes |
| 4.6.2 | completes | completes |

Three runs per cell, and two per cell in a separate environment built from scratch with an independently written test, agreeing cell for cell.

Setup: flash-attn-4 4.0.0b26, torch 2.14.0.dev20260806+cu130, CUDA 13.0. Each run in its own process under an external timeout, since the kernel never returns. Dense goes through SDPA with the FA4 impl active, bf16, B=2 H=4 S=256 D=64, non-causal. Varlen goes through `torch.nn.attention.varlen.varlen_attn`, four segments of 256, 4 heads, head dim 64. The stall is inside `_bwd_postprocess_convert`, with the same stack sampled 75 seconds apart. With grouped-query attention the varlen path stops too — 16 query heads against 8 key/value heads is enough.

Two things about how this stayed hidden, which may be more useful than the table.

The pin moved from `>=4.5.2` to `==4.6.0.dev0` in #2648 on 2026-06-25, and 4.6.0 was released on 07-02 — so no released runtime has been inside the declared range since. The pin also excludes 4.6.2, while `quack-kernels` 0.6.4 requires exactly 4.6.2, so an environment holding both cannot satisfy them.

And the version does not tell you what is loaded. `nvidia-cutlass-dsl` is a meta-package whose runtime ships in `-libs-cu12` and `-libs-cu13`; a process loads one. An environment can report 4.6.2 from `importlib.metadata` while loading a cu13 runtime still at 4.6.0 — which happens when something else pins the cu13 extra, as vllm does with `nvidia-cutlass-dsl[cu13]==4.6.0`. The reproduction below prints the mapped library for that reason.

Is 4.6.2 intended to work? If so, would you consider widening the pin?

Possibly the same thing, untested from here: #2466 (FA4 with softcap hangs on backward).

## Reproduction

```bash
pip install "flash-attn-4==4.0.0b26" --no-deps
pip install "nvidia-cutlass-dsl==4.6.0" "quack-kernels>=0.5.3" "apache-tvm-ffi>=0.1.12,<0.2"
timeout 180 python repro.py     # stops after BACKWARD_BEGIN
pip install "nvidia-cutlass-dsl==4.6.2"
timeout 180 python repro.py     # reaches BACKWARD_DONE
```

```python
import pathlib
import re

import torch
import torch.nn.attention as A
from torch.nn.attention import sdpa_kernel, SDPBackend
from importlib.metadata import version

for pkg in ("flash-attn-4", "nvidia-cutlass-dsl",
            "nvidia-cutlass-dsl-libs-cu12", "nvidia-cutlass-dsl-libs-cu13",
            "quack-kernels", "apache-tvm-ffi", "torch"):
    try:
        print(f"  {pkg} {version(pkg)}")
    except Exception:
        print(f"  {pkg} (absent)")

cap = torch.cuda.get_device_capability(0)
print(f"  gpu {torch.cuda.get_device_name(0)} sm_{cap[0]}{cap[1]}")

# Which runtime is mapped, since the version above may not say.
torch.zeros(1, device="cuda")
import cutlass  # noqa: F401
for line in pathlib.Path("/proc/self/maps").read_text().splitlines():
    if re.search(r"cute_dsl_runtime|_cutlass_ir", line):
        print(f"  loaded {line.split()[-1]}")
        break

A.activate_flash_attention_impl("FA4")
assert A.current_flash_attention_impl() == "FA4"

torch.manual_seed(0)
q, k, v = (
    torch.randn(2, 4, 256, 64, device="cuda", dtype=torch.bfloat16,
                requires_grad=True)
    for _ in range(3)
)

with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
    out = torch.nn.functional.scaled_dot_product_attention(q, k, v, is_causal=False)
torch.cuda.synchronize()
print("FORWARD_DONE", flush=True)

print("BACKWARD_BEGIN", flush=True)
out.sum().backward()
torch.cuda.synchronize()
print("BACKWARD_DONE", flush=True)

print("GRADS_FINITE",
      all(t.grad is not None and torch.isfinite(t.grad).all() for t in (q, k, v)))
```
